src/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs

```diff
@@ -793,27 +660,64 @@ private UserPreferenceCategory GetUserPreferenceCategory(int msg, IntPtr wParam,
             return pref;
         }
 
-        private void Initialize()
+        private unsafe void Initialize()
         {
             _consoleHandler = new Interop.Kernel32.ConsoleCtrlHandlerRoutine(ConsoleHandlerProc);
-            if (!Interop.Kernel32.SetConsoleCtrlHandler(_consoleHandler, true))
+            _consoleHandlerPointer = Marshal.GetFunctionPointerForDelegate(_consoleHandler);
+
+            if (!Interop.Kernel32.SetConsoleCtrlHandler(_consoleHandlerPointer, true))
             {
                 Debug.Fail("Failed to install console handler.");
                 _consoleHandler = null;
             }
 
-            _windowHandle = CreateBroadcastWindow();
-            Debug.WriteLineIf(_windowHandle == IntPtr.Zero, "CreateBroadcastWindow failed");
+            IntPtr hInstance = Interop.Kernel32.GetModuleHandle(null);
+
+            s_className = string.Format(
+                ".NET-BroadcastEventWindow.{0:x}.0",
+                AppDomain.CurrentDomain.GetHashCode());
+
+            fixed (char* c = s_className)
+            {
+                // It is important that we stash the delegate to ensure it doesn't
+                // get collected by the GC.
+
+                _windowProc = WindowProc;
+
+                Interop.User32.WNDCLASS windowClass = new Interop.User32.WNDCLASS
+                {
+                    hbrBackground = (IntPtr)(Interop.User32.COLOR_WINDOW + 1),
+                    lpfnWndProc = Marshal.GetFunctionPointerForDelegate(_windowProc),
+                    lpszClassName = c,
+                    hInstance = hInstance
+                };
 
-            AppDomain.CurrentDomain.ProcessExit += new EventHandler(SystemEvents.Shutdown);
-            AppDomain.CurrentDomain.DomainUnload += new EventHandler(SystemEvents.Shutdown);
+                if (Interop.User32.RegisterClassW(windowClass) == 0)
+                {
+                    _windowProc = null;
+                    Debug.WriteLine("Unable to register broadcast window class: {0}", Marshal.GetLastWin32Error());
+                }
+                else
+                {
+                    // And create an instance of the window.
+                    _windowHandle = Interop.User32.CreateWindowExW(
+                        0,
+                        s_className,
+                        s_className,
+                        Interop.User32.WS_POPUP,
+                        0, 0, 0, 0, IntPtr.Zero, IntPtr.Zero,
+                        hInstance, IntPtr.Zero);
```

Comment:
What happens if CreateWindowExW fails?

Reply:
We never did anything with the failed result before. Well, not until `EnsureEvents`, which sometimes explicitly does not care (see `requireHandle`).